lib/commonmarker/node.rb

```diff
@@ -0,0 +1,43 @@
+module CommonMarker
+  class Node
+    include Enumerable
+
+    # Public: An iterator that "walks the tree," descending into children recursively.
+    #
+    # blk - A {Proc} representing the action to take for each child
+    def walk(&block)
+      yield self
+      each do |child|
+        child.walk(&block)
+      end
+    end
```

Comment:
It might be nice to return an [`Enumerator`](http://ruby-doc.org/core-2.3.1/Enumerator.html#method-c-new) from `walk`, if no block is given, to match the behavior of `each`.